File: backend/app/services/campaign_service.py
from sqlalchemy.orm import Session
from ..models.campaign import Campaign, CampaignStatus
from ..services.xrpl_service import xrpl_service
from ..config import settings

class CampaignService:
    @staticmethod
    async def create_campaign(db: Session, artist_address: str, song_title: str, song_url: str, amount: float) -> Campaign:
        campaign = Campaign(
            artist_address=artist_address,
            song_title=song_title,
            song_url=song_url,
            total_amount=amount,
            amount_per_second=settings.PAYMENT_PER_SECOND,
            remaining_amount=amount,
            status=CampaignStatus.UNPAID.value
        )
        db.add(campaign)
        db.commit()
        db.refresh(campaign)
        return campaign

    @staticmethod
    async def verify_and_update_payment(db: Session, campaign_id: int, transaction_hash: str) -> Campaign:
        campaign = db.query(Campaign).filter(Campaign.id == campaign_id).first()
        if not campaign:
            raise ValueError("Campagne non trouvée")

        try:
            transaction = await xrpl_service.verify_transaction(transaction_hash)
            print("Transaction XRPL:", transaction)

            if not isinstance(transaction, dict):
                raise ValueError("Format de réponse XRPL invalide")
            
            if transaction.get("status") != "success":
                raise ValueError(transaction.get("message", "La transaction n'a pas pu être vérifiée"))
            
            amount = transaction.get("amount")
            if amount is None:
                raise ValueError("Montant de la transaction non trouvé")

            if float(amount) >= campaign.total_amount:
                campaign.status = CampaignStatus.PAID.value
                campaign.payment_transaction_hash = transaction_hash
                db.commit()
                db.refresh(campaign)
                return campaign
            else:
                raise ValueError("Le montant payé est insuffisant")

        except Exception as e:
            print(f"Erreur lors de la vérification du paiement: {str(e)}")
            raise ValueError(str(e))

    @staticmethod
    def get_active_campaigns(db: Session):
        return db.query(Campaign).filter(
            Campaign.status == CampaignStatus.PAID.value
        ).all()

    @staticmethod
    def get_artist_campaigns(db: Session, artist_address: str):
        return db.query(Campaign).filter(
            Campaign.artist_address == artist_address
        ).all()

    @staticmethod
    def delete_campaign(db: Session, campaign_id: int, artist_address: str):
        campaign = db.query(Campaign).filter(
            Campaign.id == campaign_id,
            Campaign.artist_address == artist_address,
            Campaign.status == CampaignStatus.UNPAID.value
        ).first()
        
        if not campaign:
            raise ValueError("Campagne non trouvée ou non supprimable")
            
        db.delete(campaign)
        db.commit()

campaign_service = CampaignService() 

File: frontend/src/services/campaign.service.ts
const API_URL = 'http://localhost:8000/api';

export interface Campaign {
    id: number;
    artist_address: string;
    song_title: string;
    song_url: string;
    total_amount: number;
    amount_per_second: number;
    remaining_amount: number;
    status: string;
    created_at: string;
    payment_transaction_hash?: string;
}

export interface CreateCampaignData {
    artist_address: string;
    song_title: string;
    song_url: string;
    amount: number;
}

export const CampaignService = {
    async createCampaign(data: CreateCampaignData): Promise<Campaign> {
        const response = await fetch(`${API_URL}/campaigns`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(data)
        });

        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Erreur lors de la création de la campagne');
        }

        return response.json();
    },

    async getActiveCampaigns(): Promise<Campaign[]> {
        const response = await fetch(`${API_URL}/campaigns/active`);
        
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Erreur lors de la récupération des campagnes actives');
        }

        return response.json();
    },

    async getArtistCampaigns(artistAddress: string): Promise<Campaign[]> {
        const response = await fetch(`${API_URL}/campaigns/artist/${artistAddress}`);
        
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Erreur lors de la récupération des campagnes');
        }

        return response.json();
    },

    async deleteCampaign(campaignId: number, artistAddress: string): Promise<void> {
        const response = await fetch(`${API_URL}/campaigns/${campaignId}?artist_address=${artistAddress}`, {
            method: 'DELETE'
        });

        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Erreur lors de la suppression de la campagne');
        }
    },

    async verifyPayment(campaignId: number, transactionHash: string): Promise<Campaign> {
        const response = await fetch(`${API_URL}/campaigns/${campaignId}/verify-payment`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({ transaction_hash: transactionHash })
        });

        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.detail || 'Erreur lors de la vérification du paiement');
        }

        return response.json();
    }
}; 